lib/pdn/pdn.c

```diff
@@ -0,0 +1,768 @@
+				  char *const resp,
+				  const enum pdn_at_evt evt_id)
+{
+	const char *compare_str = pdn_at_evts_list[evt_id].at_str;
+	const int compare_length = strlen(compare_str);
+
+	if (strlen(resp) >= compare_length) {
+		if (strncmp(compare_str, resp, compare_length) == 0) {
+			/* Found the event we are waiting for.
+			 * Parse parameters if any and return.
+			 */
+			return pdn_at_evts_list[evt_id].parser_fn(pdn_context,
+								  resp);
+		}
+	}
+
+	/* Invalid response received */
+	return -ENOTSUP;
+}
+
+static int command_prepare_and_send(struct pdn_connection *const pdn_context,
+				    const enum pdn_at_cmd cmd_id,
+				    const char *const remote_addr)
+{
+	int err = 0;
+	int offset = 0;
+	const int context_id =
+		(pdn_context != NULL) ? pdn_context->context_id : 0;
+	enum pdn_at_evt resp_evt = EVT_MAX;
+
+	char *const command_str = pdn_at_cmd_buffer;
+
+	if (cmd_id < CMD_MAX) {
+		/* write base if valid command, error is returned from
+		 * switch default.
+		 */
+		offset = snprintf(&command_str[offset],
+				  (PDN_AT_BUFFER_SIZE - offset), "%s",
+				  pdn_at_cmds_list[cmd_id]);
+	}
+
+	switch (cmd_id) {
+	case CMD_TIME_GET: {
+		/* No additional parameters. */
+		resp_evt = EVT_TIME_GET;
+		break;
+	}
+	case CMD_CONTEXT_CREATE: {
+		/* No additional parameters. */
+		resp_evt = EVT_CONTEXT_CREATE;
+		break;
+	}
+	case CMD_PDN_STATE_GET: {
+		/* No additional parameters. */
+		resp_evt = EVT_PDN_STATE_GET;
+		break;
+	}
+	case CMD_APN_CONNECT: {
+		/* Pack the context identifier.
+		 * Pack the IP address family string.
+		 * Pack the APN gateway string.
+		 */
+
+		offset += snprintf(&command_str[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d,",
+				   context_id);
+
+		const enum pdn_inet_type family = pdn_context->family;
+		char *family_string;
+
+		if (family == PDN_INET_IPV4V6) {
+			family_string = "IPV4V6";
+		} else if (family == PDN_INET_IPV6) {
+			family_string = "IPV6";
+		} else if (family == PDN_INET_IPV4) {
+			family_string = "IP";
+		} else {
+			return -EFAULT;
+		}
+
+		offset += snprintf(&command_str[offset],
+				   (PDN_AT_BUFFER_SIZE - offset),
+				   "\"%s\",\"%s\"", family_string, remote_addr);
+		break;
+	}
+	case CMD_PDN_ID_GET:
+		resp_evt = EVT_PDN_ID_GET;
+		/* Note: Fall through to pack the CID */
+	case CMD_PDN_ACTIVATE:
+	case CMD_PDN_DEACTIVATE:
+	case CMD_APN_DISCONNECT: {
+		/* Pack the context identifier. */
+		offset += snprintf(&command_str[offset],
+				   (PDN_AT_BUFFER_SIZE - offset), "%d",
+				   context_id);
+		break;
+	}
+	default: {
+		/* not a valid command, return error */
+		return -ENOTSUP;
+	}
+	}
+
+	if (offset > PDN_AT_BUFFER_SIZE) {
+		return -E2BIG;
+	}
+
+	/* Send AT command and wait response */
+	err = at_cmd_write(command_str, command_str, PDN_AT_BUFFER_SIZE, NULL);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Parse response. OK / ERROR is already handled by at_cmd */
+	if (resp_evt != EVT_MAX) {
+		err = command_response_parse(pdn_context, command_str,
```

Comment:
Given previous discussions, `command_response_parse` return value should not be ignored here? Currently, it's not used anywhere.

Reply:
thanks, fixed now.